propagate build results and produce a nice table

iurly opened the pull request:
* Add exception handling in futures

* Propagate build results for each built kernel, and produce a nice textual table with two results for each kernel built: one for kmod and one for ebpf.

* Also add a second table with failed kernels only, so they become easier to spot

* Set the exit code to 1 in case of failures

iurly committed:
propagate build results and produce a nice table

* Add exception handling in futures

* Propagate build results for each built kernel,
  and produce a nice textual table with two results
  for each kernel built: one for kmod and one for ebpf.

* Also add a second table with failed kernels only,
  so they become easier to spot

* Set the exit code to 1 in case of failures

gnosek commented on probe_builder/builder/distro/base_builder.py:
```diff
@@ -51,38 +77,39 @@ def build_kernel_impl(cls, config_hash, container_name, image_name, kernel_dir,
+            return cls.ProbeBuildResult(cls.ProbeBuildResult.BUILD_FAILED, took, e.output)
         else:
             took = time.time() - ts0
-            output_dir = workspace.subdir('output')
```
Why do we delete this?

iurly replied:
it got moved to the top (line 80 above) because it is needed by `probe_built()` at the very beginning to check if the probe already exists and provide a consistent `ProbeBuildResult`

gnosek replied:
Gotcha. I missed it, thanks!

gnosek commented on probe_builder/spawn.py:
```diff
@@ -20,12 +20,12 @@ def pipe(cmd, silence_errors=False, cwd=None):
         for line in lines:
             logger.debug(make_string(line))
-        return lines
+        return make_string(stdout)
```
Hmm this change looks like a no-op when combined with the one in docker.py (above). Am I missing something?

iurly replied:
could be. As a matter of fact, I find the whole `make_string()` very confusing indeed.

gnosek replied:
it's us trying to make the probe builder work both on py2 and py3, so I suppose it's obsolete by now and we could clean up some

gnosek reviewed (commented):
Minor nits, but overall it's beautiful 🤩

deepskyblue86 commented on probe_builder/builder/distro/base_builder.py:
```diff
@@ -23,6 +23,32 @@ def to_s(s):
 
 
 class DistroBuilder(object):
+    class ProbeBuildResult(object):
```
can we use enum.Enum?

deepskyblue86 commented on probe_builder/__init__.py:
```diff
@@ -137,8 +138,53 @@ def build(builder_image_prefix,
-        print("Release: {}, Result: {}".format(release, future.result()))
+        try:
+            res = future.result()
+            if res.kmod_result.failed() or res.ebpf_result.failed():
```
KernelBuildResult method?

iurly committed:
add KernelBuildResult.failed() method

iurly committed:
fix(spawn.pipe): fix return value

have spawn.pipe() return bytes() instead of a string,
so to make it consistent with subprocess.CompletedProcess().stdout

Conversely, fix default value of ProbeBuilderResult().error_log
to be bytes()

francesco-racciatti commented on probe_builder/__init__.py:
```diff
@@ -137,8 +138,53 @@ def build(builder_image_prefix,
+            print(fstr.format(release, "EXCEPTION", "EXCEPTION"))
+
+    print("-" * len(l))
+    print("Number of kernels analyzed: {}".format(len(kernels_futures)))
```
w/ python3 we can use `f""` strings

iurly replied:
Just because _you can_, it doesn't mean _you should_.
I just don't like f"" strings -- burying a variable name in a quoted string just doesn't feel right, at least to me.
Plus, they were introduced with python 3.6.
